### PR TITLE
[JENKINS-42511] Correct ComputedFolder API to include blocking behaviour for concurrent events and computations

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-42511](https://issues.jenkins-ci.org/browse/JENKINS-42511)

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -10,7 +10,7 @@
   </parent>
 
   <artifactId>cloudbees-folder</artifactId>
-  <version>5.19-SNAPSHOT</version>
+  <version>6.0.0-SNAPSHOT</version>
```
Comment (jglick) [inline pom.xml]: Whoah, that seems a little extreme for this change, which IIUC should be compatible enough.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/PeriodicFolderTrigger.java [hunk 2]
```diff
@@ -159,12 +159,13 @@ public String getInterval() {
...
+        if (job == null) {
+            return;
+        }
         FolderComputation<?> computation = job.getComputation();
-        if (computation != null) {
-            long delay = System.currentTimeMillis() - computation.getTimestamp().getTimeInMillis();
-            if (delay < interval) {
-                return;
-            }
+        long delay = System.currentTimeMillis() - computation.getTimestamp().getTimeInMillis();
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/PeriodicFolderTrigger.java]: Is this somehow related? It will produce a merge conflict with #91.

File: pom.xml [hunk 3]
```diff
@@ -19,8 +19,8 @@
   </description>
   <url>https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Folders+Plugin</url>
   <properties>
-    <jenkins.version>1.609.1</jenkins.version>
-    <java.level>6</java.level>
+    <jenkins.version>1.642.3</jenkins.version>
```
Comment (jglick) [inline pom.xml]: Why?

File: pom.xml [hunk 4]
```diff
@@ -19,8 +19,8 @@
...
+    <java.level>7</java.level>
```
Comment (jglick) [inline pom.xml]: The default; may be omitted.

File: pom.xml [hunk 5]
```diff
@@ -19,8 +19,8 @@
...
     <jenkins-test-harness.version>2.18</jenkins-test-harness.version>
```
Comment (jglick) [inline pom.xml]: Could be dropped if updating the parent.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java [hunk 6]
```diff
@@ -36,19 +36,24 @@
  * Callback for {@link ComputedFolder}.
  * Methods may be called only inside the scope of {@link ComputedFolder#computeChildren} or an out-of-band event handler.
  * @see ComputedFolder#computeChildren(ChildObserver, TaskListener)
- * @see ComputedFolder#createEventsChildObserver()
+ * @see ComputedFolder#openEventsChildObserver()
  */
-public abstract class ChildObserver<I extends TopLevelItem> {
+public abstract class ChildObserver<I extends TopLevelItem> implements Closeable {
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java]: `AutoCloseable` would suffice.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java [hunk 7]
```diff
@@ -64,12 +69,21 @@
...
+    public abstract void completed(String name);
+
+    /**
      * Returns a copy of the item names that have been observed.
      *
      * @return a copy of the item names that have been observed.
      * @since FIXME
      */
-    public abstract Set<String> observed();
+    public abstract Set<String> completed();
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java]: Why?

**Comment by jglick** [comment]
Needs merge conflict resolution BTW.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/PeriodicFolderTrigger.java (hunk 2 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/PeriodicFolderTrigger.java]: Updating the parent required this for findbugs

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java (hunk 7 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java]: Damn you IntelliJ for renaming other methods

Comment (KostyaSha) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ChildObserver.java]: Happens when initial naming was too generic :D

File: pom.xml (hunk 3 above)
Comment (stephenc) [inline pom.xml]: I want Java 7 for try with resources and this is the baseline for branch-api etc

File: pom.xml (hunk 1 above)
Comment (stephenc) [inline pom.xml]: https://github.com/jenkinsci/cloudbees-folder-plugin/pull/92/files#diff-73ee53b9a4352457ad14440d3f6692c0R282 (which I deem necessary) is the reason why I feel this is a bump to the major version.

I am using the bump to Major as a transition to a three digit scheme which will allow changes of this nature to be considered a minor bump

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolder.java [hunk 8]
```diff
@@ -119,6 +122,36 @@
     private transient FolderComputation<I> computation;
 
     /**
+     * Lock to guard {@link #currentObservations}.
+     *
+     * @since 6.0.0
+     */
+    @Nonnull
+    private transient final ReentrantLock currentObservationsLock = new ReentrantLock();
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolder.java]: [JENKINS-42593](https://issues.jenkins-ci.org/browse/JENKINS-42593)
